src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs

```diff
@@ -12,6 +12,100 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics
     [CompilerTrait(CompilerFeature.RefLocalsReturns)]
     public class RefLocalsAndReturnsTests : CompilingTestBase
     {
+        [Fact]
+        public void RefForToReadonly()
+        {
+            var comp = CreateStandardCompilation(@"
+class C
+{
+    void M(in int x)
+    {
+        for (ref int i = ref x; i < 0; i++) {}
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (6,30): error CS8329: Cannot use variable 'in int' as a ref or out value because it is a readonly variable
+                //         for (ref int i = ref x; i < 0; i++) {}
+                Diagnostic(ErrorCode.ERR_RefReadonlyNotField, "x").WithArguments("variable", "in int").WithLocation(6, 30));
+        }
+
+        [Fact]
+        public void RefForOutOfScope()
+        {
+            var comp = CreateStandardCompilation(@"
+class C
+{
+    ref int M()
+    {
+        int x = 0;
+        for (ref int i = ref x; i < 0; i++)
+        {
+            if (i == 0)
+            {
+                return ref i;
+            }
+        }
+        return ref (new int[1])[0];
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (11,28): error CS8157: Cannot return 'i' by reference because it was initialized to a value that cannot be returned by reference
+                //                 return ref i;
+                Diagnostic(ErrorCode.ERR_RefReturnNonreturnableLocal, "i").WithArguments("i").WithLocation(11, 28));
+        }
+
+        [Fact]
+        public void RefForeachReadonly()
+        {
+            var comp = CreateStandardCompilation(@"
+using System;
+class C
+{
+    void M()
+    {
+        foreach (ref var v in new int[0])
```

Comment:
We actually could make this work. 
`ref readonly` iteration for arrays is certainly doable. 
Writeable `ref` would have problem with covariance, but may still be allowed when element type is sealed.

Reply:
On the other hand an easy workaround would be - 
`foreach (ref var v in (Span<int>)new int[0])`

So maybe we should wait and see if there is a demand for `ref foreach` over arrays.
It would not be breaking to add later.

Reply:
Yup, the intent of this test was not to say that we couldn't add support for this, it was just a convenient example of an enumerator that doesn't support returning a ref.